### PR TITLE
Compose functions for URLPathArgs.

### PR DESCRIPTION
**Pull request by mzhurovich**

### CODE REVIEW

File: Blocks/URL/url.h [hunk 1]
```diff
@@ -515,6 +515,23 @@ struct URLPathArgs {
 
   void add(const std::string& arg) { args_.push_back(arg); }
 
+  std::string ComposeURLPathFromArgs() const {
+    std::string result;
+    for (auto cit = args_.crbegin(); cit != args_.crend(); ++cit) {
+      result += "/" + *cit;
```
Comment (dkorolev) [inline Blocks/URL/url.h]: `EncodeURLComponent`?

File: Blocks/URL/url.h [hunk 2]
```diff
@@ -515,6 +515,23 @@ struct URLPathArgs {
 
   void add(const std::string& arg) { args_.push_back(arg); }
 
+  std::string ComposeURLPathFromArgs() const {
+    std::string result;
+    for (auto cit = args_.crbegin(); cit != args_.crend(); ++cit) {
```
Comment (dkorolev) [inline Blocks/URL/url.h]: Won't `for (const auto& component : (*this)) {` do the trick here? It's already a reverse iterator.

File: Blocks/HTTP/test.cc [hunk 3]
```diff
@@ -171,6 +171,55 @@ TEST(HTTPAPI, RegisterWithURLPathParams) {
...
+  EXPECT_EQ("/user/a (/0, /user/a/0)", run("/user/a/0/"));
+
+  EXPECT_EQ("/user/a/1 (/, /user/a/1)", run("/user/a/1"));
+  EXPECT_EQ("/user/a/1 (/, /user/a/1)", run("/user/a/1/"));
+
+  EXPECT_EQ("/user/a (/2, /user/a/2)", run("/user/a/2"));
+  EXPECT_EQ("/user/a (/2, /user/a/2)", run("/user/a/2/"));
+
+  EXPECT_EQ("/ (/user/a/1/blah, /user/a/1/blah)", run("/user/a/1/blah"));
+  EXPECT_EQ("/ (/user/a/1/blah, /user/a/1/blah)", run("/user/a/1/blah/"));
```
Comment (dkorolev) [inline Blocks/HTTP/test.cc]: A nit from @sompylasar who happened to be looking over my shoulder: in the proper world of URLs the trailing slash, if provided originally, carries an important meaning ("is significant" as he phrased it), and is best to be kept; both in `ComposeURLPathFromArgs()` and in `ComposeURLPath()`.

I won't blame you if you decide to not give a damn as of now :-)

File: Blocks/URL/url.h (hunk 1 above)
Comment (mzhurovich) [reply Blocks/URL/url.h]: But `URLWithoutParametersParser::ComposeURL()` doesn't do it. Consistency FTW?

File: Blocks/URL/url.h (hunk 2 above)
Comment (mzhurovich) [reply Blocks/URL/url.h]: Yes, it will

File: Blocks/HTTP/test.cc (hunk 3 above)
Comment (mzhurovich) [reply Blocks/HTTP/test.cc]: `URLPathArgs` does not contain this information. It's stored separately in `Request`.

File: Blocks/URL/url.h (hunk 1 above)
Comment (dkorolev) [reply Blocks/URL/url.h]: One could argue `EncodeURLComponent` must be there for both. What if there's a slash in one of the URL path args?

File: Blocks/HTTP/test.cc (hunk 3 above)
Comment (dkorolev) [reply Blocks/HTTP/test.cc]: Ah, right. I suggest we ignore it for now then.

**Comment by dkorolev** [comment]
LGTM!

**Pull request merged**